beater/api/profile/handler.go

```diff
@@ -0,0 +1,265 @@
+
+	"github.com/elastic/apm-server/beater/headers"
+	"github.com/elastic/apm-server/beater/request"
+	"github.com/elastic/apm-server/decoder"
+	"github.com/elastic/apm-server/model/metadata"
+	"github.com/elastic/apm-server/model/profile"
+	"github.com/elastic/apm-server/publish"
+	"github.com/elastic/apm-server/transform"
+	"github.com/elastic/apm-server/utility"
+	"github.com/elastic/apm-server/validation"
+)
+
+var (
+	// MonitoringMap holds a mapping for request.IDs to monitoring counters
+	MonitoringMap = request.MonitoringMapForRegistry(registry)
+	registry      = monitoring.Default.NewRegistry("apm-server.profile", monitoring.PublishExpvar)
+)
+
+const (
+	// TODO(axw) include messageType in pprofContentType; needs fix in agent
+	pprofContentType    = "application/x-protobuf"
+	metadataContentType = "application/json"
+	requestContentType  = "multipart/form-data"
+)
+
+// Handler returns a request.Handler for managing profile requests.
+func Handler(
+	dec decoder.ReqDecoder,
+	transformConfig transform.Config,
+	report publish.Reporter,
+) request.Handler {
+	handle := func(c *request.Context) (*result, error) {
+		if c.Request.Method != http.MethodPost {
+			return nil, requestError{
+				id:  request.IDResponseErrorsMethodNotAllowed,
+				err: errors.New("only POST requests are supported"),
+			}
+		}
+		if err := validateContentType(c.Request.Header, requestContentType); err != nil {
+			return nil, requestError{
+				id:  request.IDResponseErrorsValidate,
+				err: err,
+			}
+		}
+
+		ok := c.RateLimiter == nil || c.RateLimiter.Allow()
+		if !ok {
+			return nil, requestError{
+				id:  request.IDResponseErrorsRateLimit,
+				err: errors.New("rate limit exceeded"),
+			}
+		}
+
+		// Extract metadata from the request, like user-agent and remote address.
+		reqMeta, err := dec(c.Request)
+		if err != nil {
+			return nil, requestError{
+				id:  request.IDResponseErrorsDecode,
+				err: errors.Wrap(err, "failed to decode request metadata"),
+			}
+		}
+
+		tctx := &transform.Context{
+			RequestTime: utility.RequestTime(c.Request.Context()),
+			Config:      transformConfig,
+		}
+
+		var totalLimitRemaining int64 = 10 * 1024 * 1024 // 10 MiB ought to be enough for anybody
+		var profiles []*pprof_profile.Profile
+		mr, err := c.Request.MultipartReader()
+		if err != nil {
+			return nil, err
+		}
+		for {
+			part, err := mr.NextPart()
+			if err == io.EOF {
+				break
+			} else if err != nil {
+				return nil, err
+			}
+
+			switch part.FormName() {
+			case "metadata":
+				if err := validateContentType(http.Header(part.Header), metadataContentType); err != nil {
+					return nil, requestError{
+						id:  request.IDResponseErrorsValidate,
+						err: errors.Wrap(err, "invalid metadata"),
+					}
+				}
+				r := &limitedReader{r: part, n: 10 * 1024 /* 10 KiB ought to be enough for anybody */}
+				raw, err := decoder.DecodeJSONData(r)
+				if err != nil {
+					if err, ok := r.err.(requestError); ok {
```

Comment:
Note that we're looking at `r.err` here, not `err`, where `r` is a `limitedReader`. When limitedReader identifies that the limit has been breached, it sets an error of that type.

Reply:
ofc 🤦‍♀

Reply:
Since I've moved the LimitedReader to package decoder, I've changed the way this works. I now check if the LimitedReader's N field is negative, which (as documented) means the limit was exceeded. I then construct a `requestError` at the point of checking that.